### PR TITLE
Improve sheduling of min drivers per task

### PR DESCRIPTION
**Pull request by sopel39**
Start more than single split in TaskExecutor#scheduleTaskIfNecessary.
Previously only single split was started. This caused
that less than task.min-drivers-per-task splits were
running for situations when buffer was underutlized
and no more splits were scheduled by coordinator.

### CODE REVIEW

**Comment by sopel39** [comment]
cc @dain

File: core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java [hunk 1]
```diff
@@ -384,15 +384,19 @@ private void splitFinished(PrioritizedSplitRunner split)
     private synchronized void scheduleTaskIfNecessary(TaskHandle taskHandle)
     {
         // if task has less than the minimum guaranteed splits running,
-        // immediately schedule a new split for this task.  This assures
+        // immediately schedule a new splits for this task.  This assures
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:387]:
```suggestion
        // immediately schedule new splits for this task.  This assures
```

**Review by losipiuk — COMMENTED** [review]
Code LGTM (for what I can tell :) )

Maybe make it more explicit in commit message that it is about target concurrency which is changed while execution. Maybe sht like:

```
Start more than a single split in TaskExecutor#scheduleTaskIfNecessary.
Previously only a single split was started. This caused
that less than task.min-drivers-per-task splits were
running for situations when targetConcurrency was increased in SplitConcurrencyController.
```

You mention `buffer`. What buffery you have in mind?

File: core/trino-main/src/test/java/io/trino/execution/executor/TestTaskExecutor.java [hunk 2]
```diff
@@ -443,6 +443,52 @@ public void testUserSpecifiedMaxDriversPerTask()
         }
     }
 
+    @Test
+    public void testMinDriversPerTaskWithBufferUnderflow()
```
Comment (losipiuk) [inline core/trino-main/src/test/java/io/trino/execution/executor/TestTaskExecutor.java:447]: testMinDriversPerTaskWhenTargetConcurrency increases?

File: core/trino-main/src/main/java/io/trino/util/MoreMath.java [hunk 3]
```diff
@@ -65,6 +66,13 @@ else if (a == 0 || b == 0 || diff < Float.MIN_NORMAL) {
         }
     }
 
+    public static int min(int... values)
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/util/MoreMath.java:69]: Is it needed?

**Review by arhimondr — APPROVED** [review]
The change looks good to me. The description is a little confusing though. 

My understanding is that this method is called when a split is added or when a split is finished. I guess the assumption was that it will always schedule a split if one exist. With this assumption the implementation would behave correctly.

However due to the adaptive concurrency implementation details the `pollNextSplit` method may not always return a split even if one exist, so the assumption breaks. Thus it is required to schedule as many splits as needed in one shot.

**Comment by sopel39** [comment]
> My understanding is that this method is called when a split is added or when a split is finished. I guess the assumption was that it will always schedule a split if one exist. With this assumption the implementation would behave correctly.

If worker gets N splits, but `adaptive task concurrency = N/2` only `N/2` splits will start (when all remaining driver slots are already taken by other tasks). In that case in currency code when no new splits are received, at most `N/2` splits would run until task finishes even if `adaptive task concurrency` is increased. Basically, I've observed that there happens to be a long tail when query is almost finished when query creeps forward by few splits.

**Pull request merged**